File: app/alert.rb
class Alert
  attr_reader :description, :url, :start, :end, :services

  @@date_only_length = 8  # YYYYMMDD
  @@date_time_length = 14 # YYYYMMDD HH:MM

  def initialize
    @description = ''
    @url = ''
    @start = ''
    @end = ''
    @services = []
  end

  def description_append(string)
    @description << string
  end

  def url_append(string)
    @url << string
  end

  def start_append(string)
    @start << string
  end

  def end_append(string)
    @end << string
  end

  def add_service(service)
    @services << service
  end

  def affects_trains?
    @services.any? { |service| service.type.include?('Train') }
  end

  def start_string
    # if we're in the middle of the alert time window, make the start 'Now'
    now = NSDate.date
    return 'Now' if start_nsdate <= now && end_nsdate >= now

    convert_to_string(start_nsdate, @start.length, 'Now')
  end

  def end_string
    convert_to_string(end_nsdate, @end.length, 'TBD')
  end

  private

  def start_nsdate
    @start_nsdate ||= convert_to_nsdate(@start)
  end

  def end_nsdate
    @end_nsdate ||= convert_to_nsdate(@end)
  end

  def convert_to_nsdate(time_str)
    if time_str.length == @@date_only_length
      date_only_parser.dateFromString(time_str)
    elsif time_str.length == @@date_time_length
      date_time_parser.dateFromString(time_str)
    else
      NSDate.date
    end
  end

  def convert_to_string(nsdate, str_length, str_for_empty)
    return date_time_formatter.stringFromDate(nsdate).downcase if str_length == @@date_time_length
    return date_only_formatter.stringFromDate(nsdate) if str_length == @@date_only_length
    str_for_empty
  end

  def date_time_parser
    @@date_time_parser ||= begin
      formatter = NSDateFormatter.alloc.init
      formatter.dateFormat = 'yyyyMMdd HH:mm'
      formatter.timeZone = NSTimeZone.localTimeZone
      formatter
    end
  end

  def date_time_formatter
    @@date_time_formatter ||= begin
      formatter = NSDateFormatter.alloc.init
      formatter.dateFormat = 'M/d ha'
      formatter.timeZone = NSTimeZone.localTimeZone
      formatter
    end
  end

  def date_only_parser
    @@date_only_parser ||= begin
      formatter = NSDateFormatter.alloc.init
      formatter.dateFormat = 'yyyyMMdd'
      formatter.timeZone = NSTimeZone.localTimeZone
      formatter
    end
  end

  def date_only_formatter
    @@date_only_formatter ||= begin
      formatter = NSDateFormatter.alloc.init
      formatter.dateFormat = 'M/d'
      formatter.timeZone = NSTimeZone.localTimeZone
      formatter
    end
  end
end

File: Rail Line/Alert.swift
import UIKit

class Alert {
    var description = ""
    var url = ""
    var services: [Service] = []

    private var start = ""
    private var end = ""
    private var cachedStartDate: Date?
    private var cachedEndDate: Date?

    private static let dateOnlyParser: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd"
        formatter.timeZone = NSTimeZone.local
        return formatter
    }()
    private static let dateTimeParser: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd HH:mm"
        formatter.timeZone = NSTimeZone.local
        return formatter
    }()
    private static let dateTimeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "M/d ha"
        formatter.timeZone = NSTimeZone.local
        return formatter
    }()
    private static let dateOnlyFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "M/d"
        formatter.timeZone = NSTimeZone.local
        return formatter
    }()

    func descriptionAppend(_ string: String) {
        description.append(contentsOf: string)
    }

    func urlAppend(_ string: String) {
        url.append(contentsOf: string)
    }

    func startAppend(_ string: String) {
        start.append(contentsOf: string)
    }

    func endAppend(_ string: String) {
        end.append(contentsOf: string)
    }

    func addService(_ service: Service) {
        services.append(service)
    }

    func affectsTrains() -> Bool {
        return services.contains { $0.typeIsTrain() }
    }

    func startString() -> String {
        let now = Date()
        if startDate() <= now && endDate() >= now { return "Now" }

        return dateToString(startDate(), strLength: start.count, forEmpty: "Now")
    }

    func endString() -> String {
        return dateToString(endDate(), strLength: end.count, forEmpty: "TBD")
    }

    private func convertDate(_ dateStr: String) -> Date {
        if dateStr.count == "YYYYMMDD".count {
            return Alert.dateOnlyParser.date(from: dateStr) ?? Date()
        } else if dateStr.count == "YYYYMMDD HH:MM".count {
            return Alert.dateTimeParser.date(from: dateStr) ?? Date()
        } else {
            return Date()
        }
    }

    private func dateToString(_ date: Date, strLength: Int, forEmpty: String) -> String {
        if strLength == "YYYYMMDD".count {
            return Alert.dateOnlyFormatter.string(from: date)
        } else if strLength == "YYYYMMDD HH:MM".count {
            return Alert.dateTimeFormatter.string(from: date)
        } else {
            return forEmpty
        }
    }

    private func startDate() -> Date {
        if cachedStartDate == nil { cachedStartDate = convertDate(start) }
        return cachedStartDate ?? Date()
    }

    private func endDate() -> Date {
        if cachedEndDate == nil { cachedEndDate = convertDate(end) }
        return cachedEndDate ?? Date()
    }
}
